mqtt/mqtt-broker/src/broker.rs

```diff
@@ -456,9 +472,39 @@ where
         if let Some(publication) = maybe_publication {
             self.publish_all(publication).await?
         }
+
         Ok(())
     }
 
+    async fn can_publish(
+        &mut self,
+        client_id: ClientId,
+        operation: Operation,
+    ) -> Result<bool, Error> {
+        if let Some(session) = self.sessions.get_mut(&client_id) {
+            let activity = Activity::new(session.auth_id()?.clone(), client_id.clone(), operation);
+            match self.authorizer.authorize(activity).await {
+                Ok(true) => {
+                    debug!("client {} successfully authorized", client_id);
+                    Ok(true)
+                }
+                Ok(false) => {
+                    warn!("client {} not allowed to publish", client_id);
+                    self.drop_connection(client_id).await?;
```

Comment:
I would probably drop the connection in `process_publish` not `can_publish`. I would expect a function named `can_publish` to not have any side effects.